r-rse/course-design.md

```diff
@@ -0,0 +1,135 @@
+
+- Folder and file structure (RStudio R Projects)
+    - Filepaths, very basics of the shell (in RStudio)
+    - Making use of the fs package for filesystem management
+    
+- Setting up an R package
+    - Using devtools, usethis for development
+        - e.g. `use_devtools()` and `use_usethis()`.
+    - Developing functions
+    - Making and using datasets
+
+- Function development in a package environment
+    - Dependency management
+    - Function documentation (with roxygen2), part 1
+        - To use `@examples` to help with creating function
+
+- (Idea) Additional features to make use of
+    - RStudio addins, snippets, rmarkdown templates, templates
+    - rstudioapi?
+
+- Version control: Using Git and GitHub as a sole user (part 1)
+    - Using Git in RStudio (standard add-commmit-history)
+    - Using Git in the Terminal of RStudio (moving in history with checkout,
+    creating branches, adding and updating remotes)
+    - Setting up GitHub for R package (make use of usethis), pushing and pulling
+
+- Checking correctness of code
+    - Unit test (with testthat)
+    - Code coverage (with covr)
+    - With devtools
+
+- Build management
+    - With devtools
+    - Maybe with drake?
+    - Pre-push running test, build, and check
+    - Continuous integration using Travis and/or GitHub Actions and AppVeyor 
+    (for Windows building)
+
+- Developing documentation and tutorials on usage
+    - Vignettes, README (with rmarkdown and usethis)
+    - Function documentation (with roxygen2), part 2
+    - learnr tutorials via `use_tutorials()`?
+    - Running spell checks and styling (with spelling and styler)
+
+- Involving the community:
+    - Designing your package to be used by or contributed to from the community
+    (refer to Mozilla Science Labs Open Leaders material)
+    - Contributing guidelines 
+    - Being inclusive (code of conduct)
+    - Licensing
+    - NEWS file
+
+- Exposing your package to the world with a website:
+    - Setting up a website (with pkgdown)
+    - Customizing the themes
+    - Getting all resources and material organized
+    - Defining lifecycle of your package and individual functions (with lifecycle)
+
+- Version control: Workflows around using GitHub (part 2)
+    - Make use of R builtin tools like the usethis pull request helpers 
+    (and learning how to make use of this yourself)
+
+- Project management of your package development (either sole or in a team)
+    - Issue tracking
+    - Labeling
+    - Project boards
+    - Milestones
+
+- Preparing for package release
+    - Check builds on other systems (with rhub)
+    - Running local CRAN checks
+    - Managing your versions, git tagging
+
+- Making your R package development to a team-based environment
+    - Making use of GitHub Organizations/Teams
+    - Branch protection, member roles
+    - Decision making
+    - Governance
+    - Issue assignment
+    - Relying on and using code of conduct to build culture and standards
```

Comment:
It feels like working on an R package with more than one person doesn't get covered until the end of the book, but in a class setting learners might be working together from very early on in the class.  Can we make this a different "part", and we give some guidance to instructors up front on different times pieces might be covered? E.g. you could imagine covering a classroom Code of Conduct on the first day, and get students to read about why it's a good idea to have one from this section, or when students start on their first group project, you might get them to select and agree to a Code of Conduct for their own group.